### DISCUSSION TITLE
`packageRules` Only Update `.tool-versions` for Python, Ignoring Docker and Poetry Files

### DISCUSSION THREAD

**Discussion by Dresdn** [renovatebot/renovate#28232] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

I've encountered an issue with Renovate where specifying a `packageRules` configuration to limit Python versions results in selective updates that only apply to the `.tool-versions` file, with Dockerfile and `pyproject.toml` updates being absent from the generated PRs. Without any `packageRules` specified, Renovate behaves as expected, updating Python versions across all detected dependency files including `.tool-versions`, Dockerfile, and `pyproject.toml`. However, upon introducing a `packageRules` configuration—whether through `matchDeps`, `matchPackageNames`, or `matchPackagePatterns` to restrict updates to a specific Python version range—only the `.tool-versions` file is updated in subsequent PRs.

I was expecting the packageRules to apply uniformly across all detected Python version declarations, ensuring that all dependencies, including those specified in Dockerfile and `pyproject.toml`, would be updated within the constraints defined by the `packageRules`.

Am I misunderstanding how the `packageRules` would work to manage a consistent python version across various configuration files?

**Test Repo:** https://github.com/Dresdn/renovate-python

**With `packageRules` PR**: https://github.com/Dresdn/renovate-python/pull/4

**No `packageRules` PR**: https://github.com/Dresdn/renovate-python/pull/2

### Logs (if relevant)

<details><summary>Logs</summary>

```

DEBUG: packageFiles with updates
{
  "baseBranch": "main"
  "config": {
    "asdf": [
      {
        "deps": [
          {
            "currentValue": "3.11.4",
            "depName": "python",
            "datasource": "github-tags",
            "packageName": "python/cpython",
            "extractVersion": "^v(?<version>\\S+)",
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "3.11.9",
                "newValue": "3.11.9",
                "releaseTimestamp": "2024-04-02T08:25:06.000Z",
                "newMajor": 3,
                "newMinor": 11,
                "updateType": "patch",
                "branchName": "renovate/python-3.x"
              }
            ],
            "versioning": "semver-coerced",
            "warnings": [],
            "sourceUrl": "https://github.com/python/cpython",
            "registryUrl": "https://github.com",
            "currentVersion": "3.11.4",
            "isSingleVersion": true,
            "fixedVersion": "3.11.4"
          }
        ],
        "packageFile": ".tool-versions"
      }
    ],
    "dockerfile": [
      {
        "deps": [
          {
            "depName": "docker/dockerfile",
            "currentValue": "1",
            "replaceString": "docker/dockerfile:1",
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "datasource": "docker",
            "depType": "syntax",
            "updates": [
              {
                "isPinDigest": true,
                "updateType": "pinDigest",
                "newValue": "1",
                "newDigest": "sha256:ac85f380a63b13dfcefa89046420e1781752bab202122f8f50032edf31be0021",
                "branchName": "renovate/pin-dependencies"
              }
            ],
            "packageName": "docker/dockerfile",
            "versioning": "docker",
            "warnings": [],
            "registryUrl": "https://index.docker.io",
            "currentVersion": "1",
            "fixedVersion": "1"
          },
          {
            "depName": "python",
            "currentValue": "3.11.4-alpine",
            "replaceString": "python:3.11.4-alpine",
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "datasource": "docker",
            "depType": "final",
            "updates": [
              {
                "isPinDigest": true,
                "updateType": "pinDigest",
                "newValue": "3.11.4-alpine",
                "newDigest": "sha256:603975e62d85aa07578034d3d10ffa1983b7618a6abb6371cf51941be6b8842c",
                "branchName": "renovate/pin-dependencies"
              }
            ],
            "packageName": "python",
            "versioning": "docker",
            "warnings": [],
            "registryUrl": "https://index.docker.io",
            "lookupName": "library/python",
            "currentVersion": "3.11.4",
            "fixedVersion": "3.11.4-alpine"
          }
        ],
        "packageFile": "Dockerfile"
      }
    ],
    "pep621": [
      {
        "deps": [
          {
            "packageName": "poetry-core",
            "depName": "poetry-core",
            "datasource": "pypi",
            "depType": "build-system.requires",
            "skipReason": "unspecified-version",
            "updates": []
          }
        ],
        "packageFile": "pyproject.toml"
      }
    ],
    "poetry": [
      {
        "deps": [
          {
            "datasource": "github-releases",
            "currentValue": "3.11.4",
            "managerData": {
              "nestedVersion": false
            },
            "versioning": "pep440",
            "depName": "python",
            "depType": "dependencies",
            "packageName": "containerbase/python-prebuild",
            "commitMessageTopic": "Python",
            "registryUrls": null,
            "updates": [],
            "warnings": [],
            "sourceUrl": "https://github.com/containerbase/python-prebuild",
            "registryUrl": "https://github.com",
            "currentVersion": "3.11.4",
            "fixedVersion": "3.11.4"
          }
        ],
        "packageFileVersion": "0.1.0",
        "extractedConstraints": {
          "python": "3.11.4"
        },
        "lockFiles": [
          "poetry.lock"
        ],
        "packageFile": "pyproject.toml"
      }
    ]
  }
}

```

</details>

**Comment by rarkins** [comment]
Please create two separate repos with the two different config/behavior. Also note that renovate limits itself to 2 PRs per hour by default

**Reply by Dresdn** [reply]
> Please create two separate repos with the two different config/behavior.

Done.

* https://github.com/Dresdn/renovate-python-poetry
* https://github.com/Dresdn/renovate-python-docker

> Also note that renovate limits itself to 2 PRs per hour by default

Right, but that doesn't impact updating PRs, correct? Either way, I'm ensuring that Renovate is running on the repos and not being hit by any rate limiting.

**Reply by rarkins** [reply]
Don't they have the same config? I thought you'd show working vs non working configs

**Reply by rarkins** [reply]
allowedVersions tries the configured versioning first, and if the syntax isn't supported it falls back to semver. But if the versions aren't semver compatible then it also doesn't work. 

So you could try semver-coerced versioning but you won't get pinning then

**Reply by Dresdn** [reply]
> Don't they have the same config? I thought you'd show working vs non working configs

Apologies - I misread that as different configs of each manager (Docker v Poetry) instead of Working v Not.

> So you could try semver-coerced versioning but you won't get pinning then

I created a new repo without attempting to specify a `python` version at https://github.com/Dresdn/renovate-python-non-pinned In this repo, Renovate correctly updates ASDF, Docker, and Poetry to the latest version of Python: 3.12.2

What's not making sense is that in the non-pinned repo, for each of the `"config":` types `asdf`, `docker`, and `poetry`, they all detect `"depName": "python",` and show an update to the latest python version for `"newVersion"`, and `"newValue"` is set appropriately based on the type ([as evidenced in the created PR](https://github.com/Dresdn/renovate-python-non-pinned/pull/2)).

Looking at the logs for the `non-pinned` repo run, I would expect to be able to specify `matchDepNames` and `allowedVersions`, since those are all found correctly by the default config.

Ultimately, my goal is to say, "Don't upgrade to python 12", but for the life of me, when specifying `python` under `packageNames`, Renovate isn't setting a `"newVersion"` for anything but ASDF.

**Reply by rarkins** [reply]
I'm getting a bit confused across the various repositories. Can you add a readme to the relevant one which mentions expected vs observed behavior? e.g. expected 1 PR and got 2, or expected 3 but got 1, etc?

And do you really need pin digests as part of the reproduction as in https://github.com/Dresdn/renovate-python-non-pinned/pull/1 ?

**Reply by Dresdn** [reply]
I've updated the README files for each repository:

* https://github.com/Dresdn/renovate-python -> Repo trying to limit Python to 3.11.x, but isn't working as expected (see README)
* https://github.com/Dresdn/renovate-python-non-pinned -> Basic Renovate config with no customization showing Renovate can update python for multiple depdencies (see README)

> And do you really need pin digests as part of the reproduction as in [Dresdn/renovate-python-non-pinned#1](https://github.com/Dresdn/renovate-python-non-pinned/pull/1) ?

No, but I see all the time about using a "minimal configuration" and `config:best-practices` includes pinning. I'd turn it off, but I wanted to keep the config as basic as possible.

**Reply by rarkins** [reply]
Thanks. I haven't run a debug but I suspect the problem is this:

1. Docker versioning doesn't support ranges and "less than" concepts (`<`)
2. Semver doesn't understand Docker suffixes like `-alpine`. It thinks it's an unstable version so excludes it

It could be resolved using `versionCompatibility` so that the version part is separated from the compatibility part and then the `<` syntax works.

**Reply by Dresdn** [reply]
That makes sense but `versionCompatibility` seems overkill. Looking at the debug logs for the Renovate run where nothing is configured, I feel like there's a better way to limit the version of `python` across an entire repo. For example, the `"docker"` deps shows:

```
"currentValue": "3.11.4-alpine",
"currentVersion": "3.11.4",
"datasource": "docker",
"depName": "python",
"depType": "final",
"fixedVersion": "3.11.4-alpine",
"updates": [
  {
    "bucket": "non-major",
    "newVersion": "3.12.3",
    "newValue": "3.12.3-alpine",
    "newDigest": "sha256:eb37f58646a901dc7727cf448cae36daaefaba79de33b5058dab79aa4c04aefb",
    "releaseTimestamp": "2024-04-10T00:08:34.945Z",
    "newMajor": 3,
    "newMinor": 12,
    "updateType": "minor",
    "branchName": "renovate/python-3.x"
  },
```

Based on that, I can see it extracts the version correctly, but I don't see a way to say: "For any `depName` that matches `python`, limit the `newVersion` to `3.11.x`". However, your comments seem to indicate that the [allowedVersions config option](https://docs.renovatebot.com/configuration-options/#allowedversions) matches `currentValue` instead of `currentVersion` - is that correct?

**Reply by rarkins** [reply, marked as answer]
It's not filtering on currentValue or currentVersion - it's filtering the list of tags. As an alternative you could use a regex expression for allowedVersions, e.g. `/^3\.11/` might be enough to limit it to 3.11 only

**Reply by Dresdn** [reply]
Ah - that makes sense! Switching to regex did the trick.

```
"packageRules": [
  {
    "matchDepNames": ["python"],
    "allowedVersions": "/^3\\.11/"
  }
]
```

Appreciate the help and learning!
